File: capacity_to_ship_packages_within_d_days/capacity_to_ship_packages_within_d_days.py
class Solution:
    def shipWithinDays(self, weights: List[int], D: int) -> int:
        def canShip(c): # c = capacity
            days_needed = 1
            cur_weights = 0
            
            for w in weights:
                cur_weights += w
                
                if cur_weights > c:
                    days_needed += 1
                    cur_weights = w
                    
            return days_needed <= D
    
        left = max(weights)
        right = sum(weights)
        
        while left <= right:
            mid = left + (right - left) // 2
            
            if canShip(mid):
                right = mid - 1
            else:
                left = mid + 1
        
        return left

File: capacity_to_ship_packages_within_d_days/capacity_to_ship_packages_within_d_days.js
/**
 * @param {number[]} weights
 * @param {number} days
 * @return {number}
 */
var shipWithinDays = function(weights, D) {
    function canShip(c) { // c = capacity
        let daysNeeded = 1
        let curWeights = 0

        for (const w of weights) {
            curWeights += w

            if (curWeights > c) {
                daysNeeded += 1
                curWeights = w
            }
        }

        return daysNeeded <= D
    }

    let left = Math.max(...weights)
    let right = weights.reduce((a, b) => a + b)

    while (left <= right) {
        const mid = left + Math.floor((right - left) / 2)

        if (canShip(mid))
            right = mid - 1
        else
            left = mid + 1
    }
    
    return left
};
